### PR TITLE
[Docs] Zero-to-Hero notebooks and quick start documentation

### PR DESCRIPTION
**Pull request by heyjustinai**
# What does this PR do?

Closes #

## Feature/Issue validation/testing/test plan

This PR aims to improve the documentation by creating a easy-to-follow getting started guide, as well as a zero-to-hero notebook that covers:

  - Doing Inference API Call and Fetching a Response from Endpoints
  - Creating a Conversation Loop
  - Sending Image to the Model
  - Tool Calling: How to and Details
  - Memory API: Show Simple In-Memory Retrieval
  - Agents API: Explain Components
  - Using Safety API in Conversation
  - Prompt Engineering Guide

## Please also list any relevant details for your test configuration or test plan.

n/a

## Sources

n/a

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/meta-llama/llama-stack/blob/main/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

**Comment by yanxi0830** [comment]
Thanks! FYI docs structure have been pushed, could you check that the structure/added docs are correctly reflected in https://llama-stack.readthedocs.io/en/latest/

**Review by HamidShojanazeri — COMMENTED** [review]
qq thoughts on the overall structure, wonder if instead of having these under "docs" Folder we should have a "Zero2Hero" or "Getting started" folder host these notebooks.  This would help with the user experience/ discoverability.

File: docs/zero_to_hero_guide/00_Inference101.ipynb
Comment (HamidShojanazeri) [inline docs/zero_to_hero_guide/00_Inference101.ipynb:2]: Lets add a note to make sure user know having spawned a server is assumed.

Comment (heyjustinai) [reply docs/zero_to_hero_guide/00_Inference101.ipynb:2]: I've added a note clarifying that having a server spawned is assumed, and also linked to the get started guide.

File: docs/zero_to_hero_guide/00_Inference101.ipynb
Comment (wukaixingxp) [inline docs/zero_to_hero_guide/00_Inference101.ipynb:326]: Here the UserMessage(content=response.completion_message.content, role='user') should be CompletionMessage(content=response.completion_message.content, role='assistant')?

File: docs/zero_to_hero_guide/00_Inference101.ipynb
Comment (wukaixingxp) [inline docs/zero_to_hero_guide/00_Inference101.ipynb:391]: models_response was never used in this cell, should be deleted?

File: docs/zero_to_hero_guide/02_Image_Chat101.ipynb
Comment (wukaixingxp) [inline docs/zero_to_hero_guide/02_Image_Chat101.ipynb:153]: "logo.png" did not exist anymore, use "_static/llama-stack-logo.png"?

File: docs/zero_to_hero_guide/00_Inference101.ipynb
Comment (heyjustinai) [reply docs/zero_to_hero_guide/00_Inference101.ipynb]: removed

File: docs/zero_to_hero_guide/00_Inference101.ipynb
Comment (heyjustinai) [reply docs/zero_to_hero_guide/00_Inference101.ipynb:326]: thanks for checking this, fixed.

File: docs/zero_to_hero_guide/03_Tool_Calling101.ipynb
Comment (wukaixingxp) [inline docs/zero_to_hero_guide/03_Tool_Calling101.ipynb:149]: base_url="http://{HOST}:{PORT}" should be base_url=f"http://{HOST}:{PORT}"

File: docs/zero_to_hero_guide/05_Safety101.ipynb
Comment (heyjustinai) [inline docs/zero_to_hero_guide/05_Safety101.ipynb:1]: perhaps we could show an example of how to use prompt guard as well.

**Comment by yanxi0830** [comment]
Do we expect these to be in the hosted docs site or living in the Github repo? cc @raghotham

**Comment by init27** [comment]
@yanxi0830 we were thinking the quickstart MD file should live on the docs site and point to the notebooks living inside the repo

**Comment by heyjustinai** [comment]
@yanxi0830 @init27 @raghotham It could also be both, hosting on doc and also living in repo, we could use this [library](https://github.com/jupyter/nbconvert) to convert notebooks into md

File: docs/zero_to_hero_guide/quickstart.md
Comment (subramen) [inline docs/zero_to_hero_guide/quickstart.md:22]: Consider adding code to create a new venv. Poetry with a pyproject.toml might be easier than conda, but I'll defer to you on that

File: docs/zero_to_hero_guide/quickstart.md
Comment (jeffxtang) [inline docs/zero_to_hero_guide/quickstart.md:45]: The testing below uses Llama3.1-8B-Instruct so here it may be better to use the same. Besides, running 3.2 11B vision will require more GPU.

File: docs/zero_to_hero_guide/quickstart.md
Comment (jeffxtang) [inline docs/zero_to_hero_guide/quickstart.md:162]: For each of the notebooks here, can we 
1) add an "Open in Colab" link, which can be manually but accurately generated from the github repo notebook link (cf https://github.com/meta-llama/llama-recipes/blob/main/recipes/quickstart/Prompt_Engineering_with_Llama_3.ipynb);
2) show the use of the Together's llama stack server API (in addition to locally launched server) - example https://colab.research.google.com/drive/1AmdIPJG6ZqWVaB4zUTKfcZ8AvdnWX_3J#scrollTo=_Yc4a_abBCeh, for developers who can't install a local server or prefer to have a quick start.

**Review by jeffxtang — COMMENTED** [review]
@init27 @wukaixingxp @heyjustinai

File: docs/zero_to_hero_guide/quickstart.md
Comment (subramen) [reply docs/zero_to_hero_guide/quickstart.md:162]: +1 on together

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md]: agreed! made a change to reflect this.

File: docs/zero_to_hero_guide/quickstart.md
Comment (ashwinb) [inline docs/zero_to_hero_guide/quickstart.md:8]: I'd just combine 1-2-3 into a "Setting up" section. Each of these is a small section so it feels unnecessarily too many steps (hence overwhelming on the page and to the user)

File: docs/zero_to_hero_guide/quickstart.md
Comment (ashwinb) [inline docs/zero_to_hero_guide/quickstart.md:56]: typo: `distributions`

File: docs/zero_to_hero_guide/quickstart.md
Comment (ashwinb) [inline docs/zero_to_hero_guide/quickstart.md:45]: can we use `Llama3.2-3B-Instruct` as an example? it is a smaller model to download for people? and works pretty reasonably well?

File: docs/zero_to_hero_guide/quickstart.md
Comment (ashwinb) [reply docs/zero_to_hero_guide/quickstart.md:56]: I think we could expand this into a couple sentences maybe. At least tell the user why we are using this distribution?

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:8]: agreed, consolidated it into a single section

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md]: sg, changes made

File: docs/zero_to_hero_guide/quickstart.md
Comment (ashwinb) [reply docs/zero_to_hero_guide/quickstart.md:129]: yes 100% our examples should avoid using these classes

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md]: I agree, resolved.

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md]: yep, elaborated and also mentioned about other distro options we offer.

File: docs/zero_to_hero_guide/quickstart.md
Comment (ashwinb) [inline docs/zero_to_hero_guide/quickstart.md:20]: I don't think this should have 1. 2. 3. inside at all. Just one `Setup` section and perhaps a small subsection on downloading models.

File: docs/zero_to_hero_guide/quickstart.md
Comment (ashwinb) [inline docs/zero_to_hero_guide/quickstart.md:117]: numbering is wrong, but also can we just inline all these steps into paragraphs and not make them into sections each of which is just 2 sentences? Just `## Testing with Python` is enough followed by the content directly inline separated nicely into code blocks and paras. No sub-sub-sections or numbering needed.

File: docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb
Comment (ashwinb) [inline docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb:149]: as @subramen says, we should use 
```
message = {
  "role": "user",
  "content": "..."
}
```

here

File: docs/zero_to_hero_guide/00_Inference101.ipynb
Comment (ashwinb) [inline docs/zero_to_hero_guide/00_Inference101.ipynb:334]: let's not use these `UserMessage`, `SystemMessage`, classes please. just use regular old dicts with `role=user, content=...`

File: docs/zero_to_hero_guide/00_Inference101.ipynb
Comment (ashwinb) [inline docs/zero_to_hero_guide/00_Inference101.ipynb:352]: this specific sub-header is not necessary, feels odd in the render

File: docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb
Comment (ashwinb) [inline docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb:20]: it should always be plural and without hyphens: `Prerequisites`

File: docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb
Comment (ashwinb) [inline docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb:31]: Starting from `0. ` feels odd?

**Review by ashwinb — COMMENTED** [review]
A few comments still going through

File: docs/zero_to_hero_guide/quickstart.md
Comment (HamidShojanazeri) [inline docs/zero_to_hero_guide/quickstart.md:32]: are we expecting users to install this in an already activated conda env?

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:134]: thanks for pointing this out, i agree with the typeddict syntax, changed to reflect that

File: docs/zero_to_hero_guide/quickstart.md
Comment (HamidShojanazeri) [inline docs/zero_to_hero_guide/quickstart.md:56]: I am still seeing this message "You can now run `llama stack configure meta-reference-gpu" @yanxi0830

File: docs/zero_to_hero_guide/quickstart.md
Comment (HamidShojanazeri) [inline docs/zero_to_hero_guide/quickstart.md:67]: getting this error, @heyjustinai 

```
llama stack run ./run.yaml
Using config `run.yaml`

EnvironmentNameNotFound: Could not find conda environment: llamastack-local
You can list all discoverable environments with `conda info --envs`.
```

Comment (HamidShojanazeri) [inline docs/zero_to_hero_guide/quickstart.md:67]: Also seems llama guard has been activated which seems expecting user to download before hand.

```
llama stack run ./run.yaml
Using config `run.yaml`
Resolved 13 providers
 inner-inference => meta-reference-inference
 inner-inference => meta-reference-safety
 models => __routing_table__
 inference => __autorouted__
 inner-safety => meta0
 inner-memory => meta0
 shields => __routing_table__
 safety => __autorouted__
 memory_banks => __routing_table__
 memory => __autorouted__
 agents => meta0
 telemetry => meta0
 inspect => __builtin__

Loading model `Llama3.2-3B-Instruct`
...
AssertionError: Could not find checkpoints in: /home/hamidnazeri/.llama/checkpoints/Llama-Guard-3-1B. Please download model using `llama download --model-id Llama-Guard-3-1B`
```

File: docs/zero_to_hero_guide/quickstart.md
Comment (jeffxtang) [reply docs/zero_to_hero_guide/quickstart.md:164]: Justin thanks for adding the Colab links - the branch name "docs_improvement" in the link e.g. https://colab.research.google.com/github/meta-llama/llama-stack/blob/docs_improvement/docs/zero_to_hero_guide/04_Tool_Calling101.ipynb works now, but after the merge and if the branch gets deleted, the links will be invalidate. So the links should better be `blob/main` instead of `blob/docs_improvement` even though using `main` doesn't work before the merge. @heyjustinai

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:117]: hi @ashwinb, changed the numbering, thanks for the suggestion. I’m happy to switch to the inline style. However, I believe numbering makes the steps easier to follow

File: docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb
Comment (heyjustinai) [reply docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb]: thanks, fixed this and the other occurrence on the other notebooks

File: docs/zero_to_hero_guide/00_Inference101.ipynb
Comment (heyjustinai) [reply docs/zero_to_hero_guide/00_Inference101.ipynb]: thanks, fixed this and the other occurrence on the other notebooks

File: docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb
Comment (heyjustinai) [reply docs/zero_to_hero_guide/01_Local_Cloud_Inference101.ipynb:120]: thanks for pointing this out, fixed.

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:30]: will add a note to have user activate conda prior to installing

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:20]: hi @ashwinb, same as previous comment ask well I’m happy to remove the subsections, but I believe numbering makes the steps easier to follow

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:73]: thanks for pointing it out, fixed

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:24]: will stick with conda for now for consistency

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:170]: @jeffxtang thanks for pointing this out, changed the url. I agree to have a quick start for together as well, but will create a separate notebook for that, philosophy is to keep the guide simple and easy to follow

File: docs/zero_to_hero_guide/quickstart.md
Comment (heyjustinai) [reply docs/zero_to_hero_guide/quickstart.md:62]: we are getting pypi package updated to fix this, alternatively build by source would fix it

**Review by raghotham — APPROVED** [review]
lets merge for now

**Pull request merged**
